plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,553 @@
+/*
+ * Copyright (C) 2023 Algoltek, Inc.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-algoltek-usb-common.h"
+#include "fu-algoltek-usb-device.h"
+#include "fu-algoltek-usb-firmware.h"
+#include "fu-algoltek-usb-struct.h"
+
+struct _FuAlgoltekUsbDevice {
+	FuUsbDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuAlgoltekUsbDevice, fu_algoltek_usb_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+algoltek_device_ctrl_transfer(FuAlgoltekUsbDevice *self,
+			      GUsbDeviceDirection direction,
+			      guint8 algoltek_cmd,
+			      guint16 value,
+			      guint16 index,
+			      GByteArray *transfer_pkt,
+			      guint8 len,
+			      GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   direction,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   algoltek_cmd,
+					   value,
+					   index,
+					   transfer_pkt->data,
+					   len,
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "device communication failure: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static GByteArray *
+algoltek_device_rdr(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 5);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_RDR);
+	fu_struct_algoltek_cmd_address_pkt_set_address(buf, address);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   FU_ALGOLTEK_CMD_RDR,
+					   address,
+					   0xFFFF,
+					   buf,
+					   buf->len,
+					   error))
+		return NULL;
+
+	/* success */
+	return g_steal_pointer(&buf);
+}
+
+static GByteArray *
+algoltek_device_rdv(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_transfer_pkt_new();
+	g_autoptr(GByteArray) version_data = g_byte_array_new();
+
+	fu_struct_algoltek_cmd_transfer_pkt_set_len(buf, 3);
+	fu_struct_algoltek_cmd_transfer_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_RDV);
+	fu_struct_algoltek_cmd_transfer_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   FU_ALGOLTEK_CMD_RDV,
+					   0xFFFF,
+					   0xFFFF,
+					   buf,
+					   buf->len,
+					   error))
+		return NULL;
+
+	/* Remove len and cmd bytes */
+	for (guint32 i = 2; i < buf->len; i++) {
+		if (buf->data[i] < 128)
+			fu_byte_array_append_uint8(version_data, buf->data[i]);
+	}
+	/* success */
+	return g_steal_pointer(&version_data);
+}
+
+static guint16
```

Comment:
Why make this an integer? It seems like it's Boolean data.

Reply:
This function readout the value after RDR cmd. The value is conbined by 2 bytes, so return value format should be `guint16`.

Reply:
But does it make sense to pass the value of RDR cmd up to caller or just a boolean pass/fail?  It seemed like it's used in a boolean fashion anyway and values aren't used.

Reply:
Thanks to your advice! Like you said, the values are't used. I combine the process to  `fu_algoltek_device_status_check( )` , and removed this function.